src/SignalR/common/Http.Connections/src/ConnectionEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,148 @@
+        public static IEndpointConventionBuilder MapConnectionHandler<TConnectionHandler>(this IEndpointRouteBuilder builder, string pattern, Action<HttpConnectionDispatcherOptions> configureOptions) where TConnectionHandler : ConnectionHandler
+        {
+            var options = new HttpConnectionDispatcherOptions();
+            // REVIEW: WE should consider removing this and instead just relying on the
+            // AuthorizationMiddleware
+            var attributes = typeof(TConnectionHandler).GetCustomAttributes(inherit: true);
+            foreach (var attribute in attributes.OfType<AuthorizeAttribute>())
+            {
+                options.AuthorizationData.Add(attribute);
+            }
+            configureOptions?.Invoke(options);
+
+            var conventionBuilder = builder.MapConnections(pattern, options, b =>
+            {
+                b.UseConnectionHandler<TConnectionHandler>();
+            });
+
+            conventionBuilder.Add(e =>
+            {
+                // Add all attributes on the ConnectionHandler has metadata (this will allow for things like)
+                // auth attributes and cors attributes to work seamlessly
+                foreach (var item in attributes)
+                {
+                    e.Metadata.Add(item);
+                }
+            });
+
+            return conventionBuilder;
+        }
+
+
+        /// <summary>
+        /// Maps incoming requests with the specified path to the provided connection pipeline.
+        /// </summary>
+        /// <param name="builder">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+        /// <param name="pattern">The route pattern.</param>
+        /// <param name="options">Options used to configure the connection.</param>
+        /// <param name="configure">A callback to configure the connection.</param>
+        /// <returns>An <see cref="IEndpointConventionBuilder"/> for endpoints associated with the connections.</returns>
+        public static IEndpointConventionBuilder MapConnections(this IEndpointRouteBuilder builder, string pattern, HttpConnectionDispatcherOptions options, Action<IConnectionBuilder> configure)
+        {
+            var dispatcher = builder.ServiceProvider.GetRequiredService<HttpConnectionDispatcher>();
+
+            var connectionBuilder = new ConnectionBuilder(builder.ServiceProvider);
+            configure(connectionBuilder);
+            var connectionDelegate = connectionBuilder.Build();
+
+            // REVIEW: Consider expanding the internals of the dispatcher as endpoint routes instead of
+            // using if statemants we can let the matcher handle
+
+            var conventionBuilders = new List<IEndpointConventionBuilder>();
+
+            // Build the negotiate application
+            var app = builder.CreateApplicationBuilder();
+            app.UseWebSockets();
```

Comment:
Negotiate shouldn't need websockets

Reply:
It does because of this https://github.com/aspnet/AspNetCore/blob/25f1f593783de4819a1e8e30dc09af7388cc68b3/src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs#L463